std/algorithm.d

```
@@ -1409,6 +1412,149 @@ unittest
+{
+    alias BinaryArgs = TypeTuple!(pred, "i", "a");
+
+    enum isRangeUnaryIterable(R) =
+        is(typeof(unaryFun!pred(R.init.front)));
+
+    enum isRangeBinaryIterable(R) =
+        is(typeof(binaryFun!BinaryArgs(0, R.init.front)));
+
+    enum isRangeIterable(R) =
+        isInputRange!R &&
+        (isRangeUnaryIterable!R || isRangeBinaryIterable!R);
+
+    enum isForeachUnaryIterable(R) =
+        is(typeof((R r) {
+            foreach (ref a; r)
+                cast(void)unaryFun!pred(a);
+        }));
+
+    enum isForeachBinaryIterable(R) =
+        is(typeof((R r) {
+            foreach (i, ref a; r)
+                cast(void)binaryFun!BinaryArgs(i, a);
+        }));
+
+    enum isForeachIterable(R) =
+        (!isForwardRange!R || isDynamicArray!R) &&
+        (isForeachUnaryIterable!R || isForeachBinaryIterable!R);
+
+    void each(Range)(Range r)
+    if (isRangeIterable!Range && !isForeachIterable!Range)
+    {
+        debug(each) pragma(msg, "Using while for ", Range.stringof);
+        static if (isRangeUnaryIterable!Range)
+        {
+            while (!r.empty)
+            {
+                cast(void)unaryFun!pred(r.front);
+                r.popFront();
+            }
+        }
+        else // if (isRangeBinaryIterable!Range)
+        {
+            size_t i = 0;
+            while (!r.empty)
+            {
+                cast(void)binaryFun!BinaryArgs(i, r.front);
+                r.popFront();
+                i++;
+            }
+        }
+    }
+
+    void each(Iterable)(Iterable r)
+        if (isForeachIterable!Iterable)
+    {
+        debug(each) pragma(msg, "Using foreach for ", Iterable.stringof);
+        static if (isForeachUnaryIterable!Iterable)
+        {
+            foreach (ref e; r)
+                cast(void)unaryFun!pred(e);
+        }
+        else // if (isForeachBinaryIterable!Iterable)
+        {
+            foreach (i, ref e; r)
+                cast(void)binaryFun!BinaryArgs(i, e);
```

Probably to avoid "return value is silently ignored" errors in pure code.